### PR TITLE
feature: email division assignment

### PR DESCRIPTION
**Pull request by DarinMao**
**BREAKING CHANGE: To make configuration and access easier, the divisions object has been reversed to be id:name rather than name:id. Division ID is also no longer expected to be a number, and can be any unique string.**

Example:
```yml
divisions:
  tj: TJHSST
  cool: Cool People
  open: Open
```

There are likely many bugs I did not catch, so this requires careful review. 

About the implementation:
- the new configuration is called `assignDivisions`
- `defaultDivision` is still supported, and overrides `assignDivisions`
- `assignDivisions` does nothing if email verification is turned off
  - in this case, either `defaultDivision` or the first division is the default
- `assignDivisions` is a list of rules, with the following settings:
  - `match`: the part of the email to match; can be `domain` (only the end), `email` (the entire email), `regex` (regex), or `any` (matches any email)
  - `value`: the value to match against
  - `divisions`: a list of division ids that the user is allowed to be in, with the first being the default
- rules are evaluated in the order they are configured, that is, the first rule that matches will be applied
- if a user email does not match any rules, then they are not allowed to register (403, `badCompetitionNotAllowed`)
- a user is not allowed to change their email if doing so would remove them from their division (403, `badEmailChangeDivision`)
- a user is not allowed to change their division if their email is not allowed (403, `badDivisionNotAllowed`)
- team member emails are NOT checked against the division rules

Example configuration:
```yml
assignDivisions:
  - match: domain
    value: tjhsst.edu
    divisions:
      - tj
      - open
  - match: email
    value: ginkoid@gmail.com
    divisions:
      - cool
      - open
  - match: any
    divisions:
      - open
```
- users with an email ending in `@tjhsst.edu` will be placed in the `tj` division and allowed to switch to `open`
- users with the email `ginkoid@gmail.com` will be placed in the `cool` division and allowed to switch to `open`
- all other users will be placed in `open` and not allowed to switch

Issues:
- Using `defaultDivision` only overrides `assignDivisions` for register, so if both are configured then a user may be allowed to register but not change divisions or email
- The client currently displays all divisions in the update profile box, and an error is displayed if an invalid action is taken. Perhaps it would be useful to only display the divisions a user is allowed to be in to prevent confusion. 
- CTFTime login does NOT work with `assignDivisions`, because no email is provided. However, it is unlikely that a CTF restricting divisions is using CTFTime anyway (it would most likely be a small internal CTF). 
- The email verification page can only display 401 errors when a 403 would be more appropriate for `badEmailChangeDivision`
- There are zero tests for this code
- probably more I haven't thought of

### CODE REVIEW

File: server/cache/leaderboard.js [hunk 1]
```diff
@@ -190,6 +190,7 @@ export const getScore = async ({ id }) => {
     return null
   }
   const split = redisResult.split(',')
+  console.log(split)
```
Comment (ethanwu10) [inline server/cache/leaderboard.js:193]:
```suggestion
```

File: server/util/restrict.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import config from '../../config/server'
+
+const restrictionChecks = {
+  domain: (email, value) => email.endsWith('@' + value),
+  email: (email, value) => email === value,
+  regex: (email, value) => (new RegExp(value)).test(email),
```
[the remark this answers is not in the document]
Comment (DarinMao) [reply server/util/restrict.js:6]: I'm not really familiar with all of rCTF yet, where would be the best place to do this?

**Comment by DarinMao** [comment]
This results in more database queries, particularly for email updating. I could not find a good solution. 

- the option is now called `divisionACLs`
- custom error is thrown by database on conflict
- defaultDivision removed
- ACLs precompiled

**Comment by DarinMao** [comment]
I put defaultDivision back, but it has the lower priority than divisionACLs when email verification is on. The client now sends the division with an email update and it is used to validate their division. If the client sends a fake division it is assumed to be correct and validated, and both the email and division are set on verification. This reduces the number of database queries required for these operations. 

This may cause some confusion in a situation where division is changed after updating email but before verifying it. Time will tell how intuitive this is. 

Thanks to Ethan for cleaning up the ACL checking and precompiling - wrote a bunch of untested code that all worked on the first try :PogChamp:

- look for stupid race conditions
- write unit/integration tests
- update example configs
- typescript?

**Comment by DarinMao** [comment]
Is there a better way to run these tests?

**Pull request merged**